File: solution/PokerEmFSharpTests/PokerLogicTest.fs
namespace PokerEmFSharpTests.Tests

open PokerEmCSharp
open NUnit.Framework
open FsUnit

[<TestFixture>] 
type ``Given a module that calculates the score for a hand`` ()=
    let handWithRoyalFlush =
        ({ Rank = Ten ; Suit = Clubs },
         { Rank = Jack ; Suit = Clubs },
         { Rank = Queen ; Suit = Clubs },
         { Rank = King ; Suit = Clubs },
         { Rank = Ace ; Suit = Clubs })

    let handWithStraightFlush =
        ({ Rank = Nine ; Suit = Clubs },
         { Rank = Ten ; Suit = Clubs },
         { Rank = Jack ; Suit = Clubs },
         { Rank = Queen ; Suit = Clubs },
         { Rank = King ; Suit = Clubs })

    let handWithFourOfAKind =
        ({ Rank = Nine ; Suit = Clubs },
         { Rank = Nine ; Suit = Hearts },
         { Rank = Nine ; Suit = Diamonds },
         { Rank = Nine ; Suit = Spades },
         { Rank = King ; Suit = Clubs })

    let handWithFullHouse =
        ({ Rank = Nine ; Suit = Clubs },
         { Rank = Nine ; Suit = Hearts },
         { Rank = Ten ; Suit = Clubs },
         { Rank = Ten ; Suit = Hearts },
         { Rank = Ten ; Suit = Spades })

    let handWithAceHighStraight =
        ({ Rank = Ten ; Suit = Hearts },
         { Rank = Jack ; Suit = Clubs },
         { Rank = Queen ; Suit = Clubs },
         { Rank = King ; Suit = Clubs },
         { Rank = Ace ; Suit = Clubs })

    let handWithFlush =
        ({ Rank = Ace ; Suit = Clubs },
         { Rank = Three ; Suit = Clubs },
         { Rank = Five ; Suit = Clubs },
         { Rank = Seven ; Suit = Clubs },
         { Rank = Nine ; Suit = Clubs })

    let handWithThreeOfAKind =
        ({ Rank = Nine ; Suit = Clubs },
         { Rank = Nine ; Suit = Hearts },
         { Rank = Nine ; Suit = Diamonds },
         { Rank = Queen ; Suit = Clubs },
         { Rank = King ; Suit = Clubs })

    let handWithTwoPairs =
        ({ Rank = Nine ; Suit = Clubs },
         { Rank = Nine ; Suit = Hearts },
         { Rank = Ten ; Suit = Clubs },
         { Rank = Ten ; Suit = Hearts },
         { Rank = King ; Suit = Clubs })

    let handWithPair =
        ({ Rank = Nine ; Suit = Clubs },
         { Rank = Nine ; Suit = Hearts },
         { Rank = Jack ; Suit = Clubs },
         { Rank = Queen ; Suit = Clubs },
         { Rank = King ; Suit = Clubs })

    let handWithNoCombination =
        ({ Rank = Ace ; Suit = Clubs },
         { Rank = Three ; Suit = Hearts },
         { Rank = Five ; Suit = Diamonds },
         { Rank = Seven ; Suit = Spades },
         { Rank = Nine ; Suit = Clubs })

    let sortTuple5 (a, b, c, d, e) =
        let xs = [ a; b; c; d; e ] |> List.sortBy (fun x -> x.Rank)
        (xs.[0], xs.[1], xs.[2], xs.[3], xs.[4])

    [<Test>] member test.
     ``when score is calculated for a hand with a royal flush returns RoyalFlush.`` ()=
            PokerEmCSharp.PokerLogic.GetScore (sortTuple5 handWithRoyalFlush) |> should equal RoyalFlush

    [<Test>] member test.
     ``when score is calculated for a hand with a ace-high straight do not return RoyalFlush.`` ()=
            PokerEmCSharp.PokerLogic.GetScore (sortTuple5 handWithAceHighStraight) |> should not (equal RoyalFlush)

    [<Test>] member test.
     ``when score is calculated for a hand with a straight flush returns StraightFlush.`` ()=
            PokerEmCSharp.PokerLogic.GetScore (sortTuple5 handWithStraightFlush) |> should equal StraightFlush

    [<Test>] member test.
     ``when score is calculated for a hand with a straight do not return StraightFlush.`` ()=
            PokerEmCSharp.PokerLogic.GetScore (sortTuple5 handWithAceHighStraight) |> should not (equal StraightFlush)

    [<Test>] member test.
     ``when score is calculated for a hand with four of a kind returns FourOfAKind.`` ()=
            PokerEmCSharp.PokerLogic.GetScore (sortTuple5 handWithFourOfAKind) |> should equal FourOfAKind

    [<Test>] member test.
     ``when score is calculated for a hand with a full house returns FullHouse.`` ()=
            PokerEmCSharp.PokerLogic.GetScore (sortTuple5 handWithFullHouse) |> should equal FullHouse

    [<Test>] member test.
     ``when score is calculated for a hand with a flush returns Flush.`` ()=
            PokerEmCSharp.PokerLogic.GetScore (sortTuple5 handWithFlush) |> should equal Flush

    [<Test>] member test.
     ``when score is calculated for a hand with a straight returns Straight.`` ()=
            PokerEmCSharp.PokerLogic.GetScore (sortTuple5 handWithAceHighStraight) |> should equal Straight

    [<Test>] member test.
     ``when score is calculated for a hand with three of a kind returns ThreeOfAKind.`` ()=
            PokerEmCSharp.PokerLogic.GetScore (sortTuple5 handWithThreeOfAKind) |> should equal ThreeOfAKind

    [<Test>] member test.
     ``when score is calculated for a hand with two pairs returns TwoPairs.`` ()=
            PokerEmCSharp.PokerLogic.GetScore (sortTuple5 handWithTwoPairs) |> should equal TwoPairs

    [<Test>] member test.
     ``when score is calculated for a hand with one par returns Pair.`` ()=
            PokerEmCSharp.PokerLogic.GetScore (sortTuple5 handWithPair) |> should equal Pair

    [<Test>] member test.
     ``when score is calculated for a hand without combinations returns HighCard.`` ()=
            PokerEmCSharp.PokerLogic.GetScore (sortTuple5 handWithNoCombination) |> should equal HighCard

File: solution/PokerEmCSharpTests/PokerLogicTest.cs
using PokerEmCSharp;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using System;

namespace PokerEmCSharpTests
{
    [TestClass]
    public class PokerLogicTest
    {
        private TestContext testContextInstance;

        public TestContext TestContext
        {
            get
            {
                return testContextInstance;
            }
            set
            {
                testContextInstance = value;
            }
        }

        #region Create hands

        private static Hand CreateHandWithRoyalFlush()
        {
            Hand hand = new Hand();
            hand.AddCard(new Card(Rank.Ten, Suit.Clubs));
            hand.AddCard(new Card(Rank.Jack, Suit.Clubs));
            hand.AddCard(new Card(Rank.Queen, Suit.Clubs));
            hand.AddCard(new Card(Rank.King, Suit.Clubs));
            hand.AddCard(new Card(Rank.Ace, Suit.Clubs));
            return hand;
        }

        private static Hand CreateHandWithStraightFlush()
        {
            Hand hand = new Hand();
            hand.AddCard(new Card(Rank.Nine, Suit.Clubs));
            hand.AddCard(new Card(Rank.Ten, Suit.Clubs));
            hand.AddCard(new Card(Rank.Jack, Suit.Clubs));
            hand.AddCard(new Card(Rank.Queen, Suit.Clubs));
            hand.AddCard(new Card(Rank.King, Suit.Clubs));
            return hand;
        }

        private static Hand CreateHandWithFourOfAKind()
        {
            Hand hand = new Hand();
            hand.AddCard(new Card(Rank.Nine, Suit.Clubs));
            hand.AddCard(new Card(Rank.Nine, Suit.Hearts));
            hand.AddCard(new Card(Rank.Nine, Suit.Diamonds));
            hand.AddCard(new Card(Rank.Nine, Suit.Spades));
            hand.AddCard(new Card(Rank.King, Suit.Clubs));
            return hand;
        }

        private static Hand CreateHandWithFullHouse()
        {
            Hand hand = new Hand();
            hand.AddCard(new Card(Rank.Nine, Suit.Clubs));
            hand.AddCard(new Card(Rank.Nine, Suit.Hearts));
            hand.AddCard(new Card(Rank.Ten, Suit.Clubs));
            hand.AddCard(new Card(Rank.Ten, Suit.Hearts));
            hand.AddCard(new Card(Rank.Ten, Suit.Spades));
            return hand;
        }

        private static Hand CreateHandWithAceHighStraight()
        {
            Hand hand = new Hand();
            hand.AddCard(new Card(Rank.Ten, Suit.Hearts));
            hand.AddCard(new Card(Rank.Jack, Suit.Clubs));
            hand.AddCard(new Card(Rank.Queen, Suit.Clubs));
            hand.AddCard(new Card(Rank.King, Suit.Clubs));
            hand.AddCard(new Card(Rank.Ace, Suit.Clubs));
            return hand;
        }

        private static Hand CreateHandWithFlush()
        {
            Hand hand = new Hand();
            hand.AddCard(new Card(Rank.Ace, Suit.Clubs));
            hand.AddCard(new Card(Rank.Three, Suit.Clubs));
            hand.AddCard(new Card(Rank.Five, Suit.Clubs));
            hand.AddCard(new Card(Rank.Seven, Suit.Clubs));
            hand.AddCard(new Card(Rank.Nine, Suit.Clubs));
            return hand;
        }

        private static Hand CreateHandWithThreeOfAKind()
        {
            Hand hand = new Hand();
            hand.AddCard(new Card(Rank.Nine, Suit.Clubs));
            hand.AddCard(new Card(Rank.Nine, Suit.Hearts));
            hand.AddCard(new Card(Rank.Nine, Suit.Diamonds));
            hand.AddCard(new Card(Rank.Queen, Suit.Clubs));
            hand.AddCard(new Card(Rank.King, Suit.Clubs));
            return hand;
        }

        private static Hand CreateHandWithTwoPairs()
        {
            Hand hand = new Hand();
            hand.AddCard(new Card(Rank.Nine, Suit.Clubs));
            hand.AddCard(new Card(Rank.Nine, Suit.Hearts));
            hand.AddCard(new Card(Rank.Ten, Suit.Clubs));
            hand.AddCard(new Card(Rank.Ten, Suit.Hearts));
            hand.AddCard(new Card(Rank.King, Suit.Clubs));
            return hand;
        }

        private static Hand CreateHandWithOnePair()
        {
            Hand hand = new Hand();
            hand.AddCard(new Card(Rank.Nine, Suit.Clubs));
            hand.AddCard(new Card(Rank.Nine, Suit.Hearts));
            hand.AddCard(new Card(Rank.Jack, Suit.Clubs));
            hand.AddCard(new Card(Rank.Queen, Suit.Clubs));
            hand.AddCard(new Card(Rank.King, Suit.Clubs));
            return hand;
        }

        private static Hand CreateHandWithNoCombination()
        {
            Hand hand = new Hand();
            hand.AddCard(new Card(Rank.Ace, Suit.Clubs));
            hand.AddCard(new Card(Rank.Three, Suit.Hearts));
            hand.AddCard(new Card(Rank.Five, Suit.Diamonds));
            hand.AddCard(new Card(Rank.Seven, Suit.Spades));
            hand.AddCard(new Card(Rank.Nine, Suit.Clubs));
            return hand;
        }

        #endregion

        #region Consistency

        [TestMethod]
        [ExpectedException(typeof(ArgumentException))]
        public void HandWithoutExactlyFiveCardsThrows()
        {
            var hand = new Hand();
            PokerLogic.GetScore(hand);
        }

        #endregion

        #region IsRoyalFlush

        [TestMethod]
        public void HandWithRoyalFlushReturnsRoyalFlush()
        {
            var hand = CreateHandWithRoyalFlush();
            var result = PokerLogic.GetScore(hand);
            Assert.AreEqual(result, Score.RoyalFlush);
        }

        [TestMethod]
        public void HandWithAceHighStraightDoesNotReturnRoyalFlush()
        {
            var hand = CreateHandWithAceHighStraight();
            var result = PokerLogic.GetScore(hand);
            Assert.AreNotEqual(result, Score.RoyalFlush);
        }

        #endregion

        #region IsStraightFlush

        [TestMethod]
        public void HandWithStraightFlushReturnsStraightFlush()
        {
            var hand = CreateHandWithStraightFlush();
            var result = PokerLogic.GetScore(hand);
            Assert.AreEqual(result, Score.StraightFlush);
        }

        [TestMethod]
        public void HandWithStraightDoesNotReturnStraightFlush()
        {
            var hand = CreateHandWithAceHighStraight();
            var result = PokerLogic.GetScore(hand);
            Assert.AreNotEqual(result, Score.StraightFlush);
        }

        #endregion

        #region IsFourOfAKind

        [TestMethod]
        public void HandWithFourOfAKindReturnsFourOfAKind()
        {
            var hand = CreateHandWithFourOfAKind();
            var result = PokerLogic.GetScore(hand);
            Assert.AreEqual(result, Score.FourOfAKind);
        }

        #endregion

        #region IsFullHouse

        [TestMethod]
        public void HandWithFullHouseReturnsFullHouse()
        {
            var hand = CreateHandWithFullHouse();
            var result = PokerLogic.GetScore(hand);
            Assert.AreEqual(result, Score.FullHouse);
        }

        #endregion

        #region IsFlush

        [TestMethod]
        public void HandWithFlushReturnsFlush()
        {
            var hand = CreateHandWithFlush();
            var result = PokerLogic.GetScore(hand);
            Assert.AreEqual(result, Score.Flush);
        }

        #endregion

        #region IsStraight

        [TestMethod]
        public void HandWithStraightReturnsStraight()
        {
            var hand = CreateHandWithAceHighStraight();
            var result = PokerLogic.GetScore(hand);
            Assert.AreEqual(result, Score.Straight);
        }

        #endregion

        #region IsThreeOfAKind

        [TestMethod]
        public void HandWithThreeOfAKindReturnsThreeOfAKind()
        {
            var hand = CreateHandWithThreeOfAKind();
            var result = PokerLogic.GetScore(hand);
            Assert.AreEqual(result, Score.ThreeOfAKind);
        }

        #endregion

        #region IsTwoPairs

        [TestMethod]
        public void HandWithTwoPairsReturnsTwoPairs()
        {
            var hand = CreateHandWithTwoPairs();
            var result = PokerLogic.GetScore(hand);
            Assert.AreEqual(result, Score.TwoPair);
        }

        #endregion

        #region IsPair

        [TestMethod]
        public void HandWithOnePairReturnsPair()
        {
            var hand = CreateHandWithOnePair();
            var result = PokerLogic.GetScore(hand);
            Assert.AreEqual(result, Score.Pair);
        }

        #endregion

        #region IsHighCard

        [TestMethod]
        public void HandWithNoCombinationReturnsHighCard()
        {
            var hand = CreateHandWithNoCombination();
            var result = PokerLogic.GetScore(hand);
            Assert.AreEqual(result, Score.HighCard);
        }

        #endregion
    }
}
